File: wear/src/main/java/org/dwallach/xstopwatch/TimePickerFragment.java
package org.dwallach.xstopwatch;

import android.app.Activity;
import android.app.DialogFragment;
import android.net.Uri;
import android.os.Bundle;
import android.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.NumberPicker;


public class TimePickerFragment extends DialogFragment {
    private int hours, minutes;

    public final static String HOURS_PARAM = "hours";
    public final static String MINUTES_PARAM = "minutes";

    /**
     * This is a totally cheesy time picker. We'd rather use the system one, but it doesn't work.
     */
    public static TimePickerFragment newInstance() {
        TimePickerFragment fragment = new TimePickerFragment();
        Bundle args = new Bundle();
        long duration = TimerState.getSingleton().getDuration(); // in milliseconds
        int minutes = (int) ((duration / 60000) % 60);
        int hours = (int) (duration / 3600000);
        args.putInt(HOURS_PARAM, hours);
        args.putInt(MINUTES_PARAM, minutes);
        fragment.setArguments(args);
        return fragment;
    }

    public TimePickerFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View mainView = inflater.inflate(R.layout.fragment_time_picker, container, false);

        if (getArguments() != null) {
            hours = getArguments().getInt(HOURS_PARAM);
            minutes = getArguments().getInt(MINUTES_PARAM);
        }
        ImageButton okButton = (ImageButton) mainView.findViewById(R.id.pickerOkButton);
        ImageButton resetButton = (ImageButton) mainView.findViewById(R.id.pickerResetButton);
        final NumberPicker hoursPicker = (NumberPicker) mainView.findViewById(R.id.hoursPicker);
        final NumberPicker minutesPicker = (NumberPicker) mainView.findViewById(R.id.minutesPicker);

        hoursPicker.setMinValue(0);
        hoursPicker.setMaxValue(23);
        hoursPicker.setWrapSelectorWheel(false);
        hoursPicker.setValue(hours);
        minutesPicker.setMinValue(0);
        minutesPicker.setMaxValue(59);
        minutesPicker.setValue(minutes);
        minutesPicker.setWrapSelectorWheel(false);
        minutesPicker.setFormatter(new NumberPicker.Formatter() {
            @Override
            public String format(int value) {
                return String.format("%02d", value);   // numbers less than ten will have a leading zero
            }
        });

        okButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                hours = hoursPicker.getValue();
                minutes = minutesPicker.getValue();
                TimerState timerState = TimerState.getSingleton();
                timerState.setDuration(null, hours * 3600000 + minutes * 60000);

                // okay, we're done!
                dismiss();
            }
        });

        resetButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                dismiss();
            }
        });

        return mainView;
    }
}

File: wear/src/main/kotlin/org/dwallach/xstopwatch/TimePickerFragment.kt
package org.dwallach.xstopwatch

import android.app.DialogFragment
import android.graphics.Paint
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageButton
import android.widget.NumberPicker
import org.jetbrains.anko.find

class TimePickerFragment : DialogFragment() {
    private var hours: Int
    private var minutes: Int

    /**
     * This is a totally cheesy time picker. We'd rather use the system one, but it doesn't work.
     */
    init {
        val duration = TimerState.duration // in milliseconds
        minutes = (duration / 60000 % 60).toInt()
        hours = (duration / 3600000).toInt()
//        arguments = Bundle().apply {
//            putInt(HOURS_PARAM, hours)
//            putInt(MINUTES_PARAM, minutes)
//        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        if(inflater == null) {
            Log.e(TAG, "no inflater, can't do anything!")
            return null
        }

        // nuke the frame around the fragment
        setStyle(DialogFragment.STYLE_NO_FRAME, 0)

        // Inflate the layout for this fragment
        val mainView = inflater.inflate(R.layout.fragment_time_picker, container, false)

        mainView.apply {
            val pickerOkButton = find<ImageButton>(R.id.pickerOkButton)
            val pickerResetButton = find<ImageButton>(R.id.pickerResetButton)
            val hoursPicker = find<NumberPicker>(R.id.hoursPicker)
            val minutesPicker = find<NumberPicker>(R.id.minutesPicker)

            hoursPicker.apply {
                setTextColor(-1)
                minValue = 0
                maxValue = 23
                wrapSelectorWheel = false
                value = hours
            }

            minutesPicker.apply {
                setTextColor(-1)
                minValue = 0
                maxValue = 59
                value = minutes
                wrapSelectorWheel = false
                setFormatter { "%02d".format(it) } // numbers less than ten will have a leading zero
            }

            pickerOkButton.setOnClickListener {
                hours = hoursPicker.value
                minutes = minutesPicker.value
                TimerState.setDuration(null, (hours * 3600000 + minutes * 60000).toLong()) // also resets the timer
                PreferencesHelper.savePreferences(context)
                PreferencesHelper.broadcastPreferences(context, Constants.timerUpdateIntent)

                // okay, we're done!
                dismiss()
            }

            pickerResetButton.setOnClickListener { dismiss() }
        }

        return mainView
    }

    companion object {
        private const val TAG = "TimePickerFragment"
//        const val HOURS_PARAM = "hours"
//        const val MINUTES_PARAM = "minutes"
    }
}

// this solution adapted from: http://stackoverflow.com/questions/18120840/numberpicker-textcolour
fun NumberPicker.setTextColor(color: Int) {
    val TAG = "NumberPicker.setTextColor"
    Log.v(TAG, "setting number picker color")

    val count = childCount
    for (i in 0..count - 1) {
        val child = getChildAt(i)
        if (child is EditText) {
            try {
                Log.v(TAG, "found an edit text field ($i)")
                val selectorWheelPaintField = javaClass.getDeclaredField("mSelectorWheelPaint")
                selectorWheelPaintField.isAccessible = true
                (selectorWheelPaintField.get(this) as Paint).color = color
                val oldColor = child.currentTextColor
                Log.v(TAG, "oldColor(%x), newColor(%x)".format(oldColor, color))
                child.setTextColor(color)
                invalidate()
                return
            } catch (e: NoSuchFieldException) {
                Log.w(TAG, e)
            } catch (e: IllegalAccessException) {
                Log.w(TAG, e)
            } catch (e: IllegalArgumentException) {
                Log.w(TAG, e)
            }
        }
    }
}
